### PR TITLE
Add module for data handling and check integrity of turbine data

### PR DESCRIPTION
**Pull request by uvchik**
1. Add function to check the integrity of the turbine data base. This could help to avoid #100 in the future.
2. Restructure the data handling and move all related functions into the new data module.

UPDATE:
3. `get_turbine_types()` will not silently import new database from oedb. In the past the local data has been overwritten by this function.

### CODE REVIEW

**Comment by uvchik** [comment]
There are still some tests missing with a wrong dataset.

At the moment the test will log suspicious data sets but will not remove them.
If the check leads to an error the data set will not be stored. This could be changed in the future, so that only broken datasets are remove but the rest will be stored.

It is possible to add additional checks.

@birgits, @SabineHaas Please check whether you agree with the basic approach.

File: windpowerlib/wind_turbine.py [hunk 1]
```diff
@@ -409,215 +406,10 @@ def get_turbine_data_from_file(turbine_type, path):
...
-    else:
-        curves_df = df[
-            ["manufacturer", "turbine_type", "has_power_curve", "has_cp_curve"]
-        ]
-    if print_out:
-        pd.set_option("display.max_rows", len(curves_df))
-        print(curves_df)
-        pd.reset_option("display.max_rows")
-    return curves_df
+    print(turbine_library, print_out, filter_)
```
Comment (birgits) [inline windpowerlib/wind_turbine.py:410]: Is this print intended?

File: windpowerlib/data.py [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+            copyfile(
+                filename.format("{0}s_{1}".format(curve_type, time_stamp)),
+                filename.format("{}s".format(curve_type)),
+            )
+        remove_tmp_file(filename, time_stamp)
+        raise e
+    return data
+
+
+def check_data_integretiy(data, min_pc_length=5):
```
Comment (birgits) [inline windpowerlib/data.py:294]: spelling error in integrity

File: windpowerlib/data.py [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+                wt.power_coefficient_curve is None
+                and dataset[1].has_cp_curve is True
+            ):
+                logging.warning(
+                    "{0}: No cp-curve but has_cp_curve=True.".format(ttype)
+                )
+            if wt.power_curve is not None:
+                if len(wt.power_curve) < min_pc_length:
+                    logging.warning(
+                        "{0}: power_curve is to short ({1} values),".format(
```
Comment (birgits) [inline windpowerlib/data.py:317]: spelling error in "too short"

File: windpowerlib/data.py [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+            )
+        remove_tmp_file(filename, time_stamp)
+        raise e
+    return data
+
+
+def check_data_integretiy(data, min_pc_length=5):
+    for dataset in data.iterrows():
+        ttype = dataset[0]
+        enercon_e126 = {"turbine_type": "{0}".format(ttype), "hub_height": 135}
```
Comment (birgits) [inline windpowerlib/data.py:297]: I would choose more generic variable name, such as wt_dict oder something

File: tests/test_wind_turbine.py [hunk 5]
```diff
@@ -48,13 +48,15 @@ def test_get_turbine_types(self, capsys):
         with pytest.raises(ValueError, match=msg):
             get_turbine_types("wrong")
 
+    def test_store_turbine_data_from_oedb(self):
+        store_turbine_data_from_oedb()
+
```
Comment (birgits) [inline tests/test_wind_turbine.py:53]: This doesn't test anything, does it?

**Comment by birgits** [comment]
Thank you @uvchik! I only had a quick look so far and made some minor comments. As far as I can see without running the code myself, it looks good. If you want a more thorough feedback it unfortunately has to wait until after my holidays.

File: windpowerlib/wind_turbine.py (hunk 1 above)
Comment (uvchik) [reply windpowerlib/wind_turbine.py:410]: For me it is fine like this.

File: tests/test_wind_turbine.py (hunk 5 above)
Comment (uvchik) [reply tests/test_wind_turbine.py:53]: I did not change this test, I just moved it. It is a pure function test. 

Now I added a test that checks if the files have been updated by the function call.

File: tests/test_data_handling.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+        self.df.loc["GE100/2750", "has_cp_curve"] = True
+        check_data_integrity(self.df, min_pc_length=26)
+        assert "E48/800: power_curve is too short (25 values)" in caplog.text
+        assert "GE158/4800: No power curve" in caplog.text
+        assert "GE100/2750: No cp-curve but has_cp_curve" in caplog.text
+
+    def test_global_error(self):
+        msg = "Must have equal len keys"
+        with pytest.raises(ValueError, match=msg):
+            self.df.loc["GE158/4800", "has_cp_curve"] = [5, 3]
```
Comment (SabineHaas) [inline tests/test_data_handling.py:73]: The ValueError is already risen in this line:
`self.df.loc["GE158/4800", "has_cp_curve"] = [5, 3]`
I couldn't fix the test as I don't fully understand what should be tested.

File: tests/test_data_handling.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    def test_get_turbine_types(self, capsys):
+        get_turbine_types()
+        captured = capsys.readouterr()
+        assert "Enercon" in captured.out
+        get_turbine_types("oedb", print_out=False, filter_=False)
+        msg = "`turbine_library` is 'wrong' but must be 'local' or 'oedb'."
+        with pytest.raises(ValueError, match=msg):
+            get_turbine_types("wrong")
+
+    def test_store_turbine_data_from_oedb(self):
```
Comment (SabineHaas) [inline tests/test_data_handling.py:92]: This test fails for me locally but works in the remote tests here. 
Error msg: `AssertionError: assert 1592378146.292112 < 1592378146.292112`

What exactly do you test here?

**Review by SabineHaas — DISMISSED** [review]
Thank you @uvchik and excuse my late review! 

I already give my approval although you should check my suggestions and comments.
I would also appreciate if you could add some more comments and docstrings - at least a one line explanation for each new function. :)

**Comment by uvchik** [comment]
A new thought about this issue. We could store the update of the turbine data at a new place. This will have some advantages:

* the user do not need the right to change the package
* the original data set will persist and it is always possible to go back to the original data set
* it is easier to add/change data without having the risk of making the original data file unreadable

File: windpowerlib/data.py [hunk 8]
```diff
@@ -0,0 +1,347 @@
...
+        Turbine data of different turbines such as 'manufacturer',
+        'turbine_type', 'nominal_power'.
+
+    """
+    turbine_data = fetch_turbine_data_from_oedb(schema=schema, table=table)
+    # standard file name for saving data
+    filename = os.path.join(os.path.dirname(__file__), "oedb", "{0}.csv")
+
+    # get all power (coefficient) curves and save to file
+    # for curve_type in ['power_curve', 'power_coefficient_curve']:
```
Comment (SabineHaas) [inline windpowerlib/data.py:189]:
```suggestion
```

**Review by SabineHaas — APPROVED** [review]
Looks good to me @uvchik !

File: windpowerlib/data.py (hunk 8 above)
Comment (uvchik) [reply windpowerlib/data.py]: Removed. Thanks :smile:

**Pull request merged**